platform-sdk/swirlds-platform-core/src/main/java/com/swirlds/platform/AddressBookInitializer.java

```diff
@@ -0,0 +1,254 @@
+    /** The address book loaded from the saved state on disk. May be null. */
+    private final AddressBook stateAddressBook;
+    /** The address book determined for use after {@link AddressBookInitializer#initialize()} is called. */
+    private final AddressBook initialAddressBook;
+    /** The path to the directory for writing address books. */
+    private final Path pathToAddressBookDirectory;
+    /** Indicate that the unmodified config address book must be used. */
+    private final boolean useConfigAddressBook;
+
+    /**
+     * Constructs an AddressBookInitializer to initialize an address book from the swirld application state, config.txt,
+     * and the saved state from disk.
+     *
+     * @param currentVersion       The current version of the application. Must not be null.
+     * @param signedState          The signed state loaded from disk.  May be null.
+     * @param genesisSwirldState   The swirld application state in genesis start. Must not be null.
+     * @param configAddressBook    The address book derived from config.txt. Must not be null.
+     * @param useConfigAddressBook Indicates if the unmodified config address book should be used.
+     */
+    public AddressBookInitializer(
+            final SoftwareVersion currentVersion,
+            final SignedState signedState,
+            final SwirldState genesisSwirldState,
+            final AddressBook configAddressBook,
+            final String parentDirectory,
+            final boolean useConfigAddressBook) {
+        this.currentVersion = Objects.requireNonNull(currentVersion, "The currentVersion must not be null.");
+        this.genesisSwirldState = Objects.requireNonNull(genesisSwirldState, "The swirldState must not be null.");
+        this.configAddressBook = Objects.requireNonNull(configAddressBook, "The configAddressBook must not be null.");
+        this.useConfigAddressBook = useConfigAddressBook;
+        Path addressBookDirectoryPath = Path.of(parentDirectory, ADDRESS_BOOK_DIRECTORY_NAME);
+        try {
+            Files.createDirectories(addressBookDirectoryPath);
+        } catch (IOException e) {
+            logger.error(EXCEPTION.getMarker(), "Not able to create directory: {}", addressBookDirectoryPath, e);
+            addressBookDirectoryPath = null;
+        }
+        this.pathToAddressBookDirectory = addressBookDirectoryPath;
+
+        if (signedState != null) {
+            this.stateAddressBook = signedState.getAddressBook();
+            this.stateSwirldState = signedState.getSwirldState();
+            SoftwareVersion stateVersion;
+            try {
+                stateVersion = signedState
+                        .getState()
+                        .getPlatformState()
+                        .getPlatformData()
+                        .getCreationSoftwareVersion();
+            } catch (Exception e) {
+                logger.error(EXCEPTION.getMarker(), "Unable to retrieve software version from signed state.");
```

Comment:
it seems like we should not proceed if this happens

Reply:
Is there a guarantee that if the signedState is not null that all parts of the data substructure are accessible?

Reply:
I don't think there is a guarantee. But if we expect it to not be null, and it is null, we should not proceed with the startup

Reply:
If we are reading a state from disk (i.e. not starting at genesis), we do not permit anything of these things to be null. There are many things that will break elsewhere if that is the case.

The non-bug only pathway that could possibly introduce nulls into these variables would be to tamper with the signed state file by hand (which is not something I think we should bother trying to support).

Reply:
I've collapsed the code paths to only be sensitive to whether or not the loadedSignedState is null or non-null.  If non-null, I assume that all parts of the state are non-null.